### PR TITLE
feat(validation): validate concerto objects from dotnet using node

### LINKED ISSUE #5: Port Validation from JavaScript Implementation
# Feature Request 🛍️
To mirror the functional API for Concerto, we need to implement validation of a JSON object against a model definition.

## Use Case

Validation is a key operation of a schema language, and is used to determine whether data is valid with respect to the model and the language specification.

## Possible Solution

The existing (empty) `AccordProject.Concerto` namespace should have a public class added that contains a `validate` operation with the following signature:

```cs
    public void validate(String jsonObject, ObjectValidatorOptions options) {
        //...
    }
```

Note that this is not a binary operation, i.e. there is no `true`/`false` response depending on the outcome of validation. Instead, a validation of a valid object will return successfully and validation of an invalid object will throw an exception that contains details of why the validation failed.

## Context

The reference implementation for this is in the JS version of Concerto.

- The public API signature, https://github.com/accordproject/concerto/blob/master/packages/concerto-core/lib/concerto.js#L51
- The Object Validator visitor, https://github.com/accordproject/concerto/blob/master/packages/concerto-core/lib/serializer/objectvalidator.js

It is critical that implementations of Concerto for different runtimes have the same semantics and very close signatures. Extensions to the language should be avoided until parity is achieved with the JS implementation. It is expected that the JS implementation will continue to be the reference implementation for some time. 

## Detailed Description

A proper implementation of validation will also require porting of the ModelManager and associated introspection Classes (see #4).

Existing unit tests, which should be ported
- [Basic](https://github.com/accordproject/concerto/blob/master/packages/concerto-core/test/serializer/objectvalidator.js) 
- [Complex](https://github.com/accordproject/concerto/blob/master/packages/concerto-core/test/serializer/complex/complex.js) (serialization implicitly includes validation), 
- [Functional API](https://github.com/accordproject/concerto/blob/master/packages/concerto-core/test/1.0.0/validate.js)

Dates are specified according to the ISO 8601 standard. We adopt the DayJS library which supports the following formats This spec must be respected in the .NET port. All of the following are valid date formats in Concerto.

```
YYYY-MM-DDTHH:MM:SS.000+hh:mm
YYYY-MM-DDTHH:MM:SS.000+hhmm
YYYY-MM-DDTHH:MM:SS.000Z
YYYY-MM-DDTHH:MM:SS.000
YYYY-MM-DDTHH:MM:SS
YYYY-MM-DDTHH:MM:SSZ
YYYY-MM-DD
YYYY-MM
YYYYMMDD
YYYYMMDDTHHMMSS
```
 
 Regular expressions use the ECMAScript flavour, any validation of Regular Expressions in field validators should be performed in [ECMAScript compatibility mode](https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#ecmascript-matching-behavior).
 
Error messages and console logs must not include any values from the input data, e.g. https://github.com/accordproject/concerto/pull/483

### PR DESCRIPTION
**Pull request by jamieshorten**
# Closes #5

Validate Concerto objects against a provided concerto models from .Net code using [Javascript.NodeJS](https://github.com/JeringTech/Javascript.NodeJS) to run the core Concerto engine in NodeJS.

### Changes
This works by wrapping the core Concerto engine with a TypeScript function that is then webpacked to Module that Node can execute. [Javascript.NodeJS](https://github.com/JeringTech/Javascript.NodeJS) takes care of orchestrating the communication between the host .Net application and NodeJS.

#### How Javascript.NodeJS works
This package works by serialising any parameters using the built in dotnet JSON serialiser. It then calls a function in a NodeJS module with the supplied parameters. The module needs to contain functions that conform to a signature defined [here](https://github.com/JeringTech/Javascript.NodeJS#invoking-javascript).

The module can either be passed as a string or a file path. In the case of a file path the module is cached for as long as the NodeJS process is running. If the module is passed as a string then the client needs to specify a cache key manually if it needs the module to be cached.

The NodeJS processes spawned by Javascript.NodeJS are terminated when the `INodeJSService` is disposed.

Returned objects are deserialised to a client provided type.
All communication with the NodeJS process happens over HTTP.

Finally this project does not handle creation of the `INodeJSService` instance. The client application should do this and use whatever DI system it uses to inject it into the `AccordProject.Concerto.Validate.Validator` instance at construction time. This is to allow the client to configure any specific NodeJS settings it needs such as [multiprocess concurrency](https://github.com/JeringTech/Javascript.NodeJS#enabling-multi-process-concurrency).

### Flags

NodeJS needs to be present on the host system. There are two major alternatives to this approach:
 - [ClearScript](https://github.com/microsoft/ClearScript), which embeds V8.
 - One of the other embeddable JS engines (some are even written entirely in .Net)

Neither approach is suitable because Concerto requires enough functionality that is only present in either the browser or NodeJS that shimming or polyfilling the required types and functions is impractical.

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `master` from `fork:branchname`

### CODE REVIEW

**Comment by dselman** [comment]
It would be wonderful if you could describe the processes/architecture for a .NET application that is using this. For example, how does the  data get serialised between .NET and Node?

**Comment by jamieshorten** [comment]
> It would be wonderful if you could describe the processes/architecture for a .NET application that is using this. For example, how does the data get serialised between .NET and Node?

Added a brief description

File: AccordProject.Concerto.Validate/concerto-validate/dist/validate.js
Comment (mttrbrts) [inline AccordProject.Concerto.Validate/concerto-validate/dist/validate.js:1]: The dist folder should be ignored and created as part of the build process.

**Review by mttrbrts — DISMISSED** [review]
Approved, with one minor comment

File: AccordProject.Concerto.Validate/concerto-validate/dist/validate.js
Comment (jamieshorten) [reply AccordProject.Concerto.Validate/concerto-validate/dist/validate.js:1]: Addressed. `dist/*` is now ignored.

**Review by mttrbrts — DISMISSED** [review]
Please merge once the build and DCO are fixed.

**Review by mttrbrts — APPROVED** [review]

**Pull request merged**